File: Common/Json/Dynamic/JsonDynamicValue.vb
Namespace Common

    Public Class JsonDynamicValue
        Inherits JsonDynamicBase

        Public Sub New(ByVal Value As String, ByVal IsString As Boolean)
            Me._Value = Value
            Me._IsString = IsString
        End Sub

        Public Shared Widening Operator CType(ByVal Value As String) As JsonDynamicValue
            Return New JsonDynamicValue(Value, True)
        End Operator

        Public Shared Widening Operator CType(ByVal Value As JsonDynamicValue) As String
            Verify.True(Value.IsString, "Value is not a string.")
            Return Value.Value
        End Operator

        Public Shared Widening Operator CType(ByVal Value As Integer) As JsonDynamicValue
            Return New JsonDynamicValue(Value.ToStringInv(), False)
        End Operator

        Public Shared Widening Operator CType(ByVal Value As JsonDynamicValue) As Integer
            Verify.False(Value.IsString, "Value is a string.")
            Return Value.GetInteger()
        End Operator

        Public Shared Widening Operator CType(ByVal Value As Double) As JsonDynamicValue
            Return New JsonDynamicValue(Value.ToStringInv(), False)
        End Operator

        Public Shared Widening Operator CType(ByVal Value As JsonDynamicValue) As Double
            Verify.False(Value.IsString, "Value is a string.")
            Return Value.GetDouble()
        End Operator

        Public Shared Widening Operator CType(ByVal Value As Boolean) As JsonDynamicValue
            Return New JsonDynamicValue(If(Value, "true", "false"), False)
        End Operator

        Public Shared Widening Operator CType(ByVal Value As JsonDynamicValue) As Boolean
            Verify.False(Value.IsString, "Value is a string.")
            Return Value.GetBoolean()
        End Operator

        Public Shared Operator =(ByVal L As JsonDynamicValue, ByVal R As JsonDynamicValue) As Boolean
            Return L.Value = R.Value And L.IsString = R.IsString
        End Operator

        Public Shared Operator <>(ByVal L As JsonDynamicValue, ByVal R As JsonDynamicValue) As Boolean
            Return Not L = R
        End Operator

        Public Overrides Function Equals(obj As Object) As Boolean
            Dim T = TryCast(obj, JsonDynamicValue)
            If T Is Nothing Then
                Return False
            End If
            Return Me = T
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return Utilities.CombineHashCodes(Me.Value.GetHashCode(), Me.IsString.GetHashCode())
        End Function

        Public Overrides Function ToString() As String
            Return If(Me.IsString, """", "") & Me.Value & If(Me.IsString, """", "")
        End Function

#Region "Value Read-Only Property"
        Private ReadOnly _Value As String

        Public ReadOnly Property Value As String
            Get
                Return Me._Value
            End Get
        End Property
#End Region

#Region "IsString Read-Only Property"
        Private ReadOnly _IsString As Boolean

        Public ReadOnly Property IsString As Boolean
            Get
                Return Me._IsString
            End Get
        End Property
#End Region

        Friend Const [True] = "true"
        Friend Const [False] = "false"

    End Class

End Namespace

File: Common/Json/Dynamic/JsonDynamicValue.cs
namespace Ks.Common
{
        public class JsonDynamicValue : JsonDynamicBase
        {
            public JsonDynamicValue(string Value, bool IsString)
            {
                this._Value = Value;
                this._IsString = IsString;
            }

            public static implicit operator JsonDynamicValue(string Value)
            {
                return new JsonDynamicValue(Value, true);
            }

            public static implicit operator string(JsonDynamicValue Value)
            {
                Verify.True(Value.IsString, "Value is not a string.");
                return Value.Value;
            }

            public static implicit operator JsonDynamicValue(int Value)
            {
                return new JsonDynamicValue(Value.ToStringInv(), false);
            }

            public static implicit operator int(JsonDynamicValue Value)
            {
                Verify.False(Value.IsString, "Value is a string.");
                return Value.GetInteger();
            }

            public static implicit operator JsonDynamicValue(double Value)
            {
                return new JsonDynamicValue(Value.ToStringInv(), false);
            }

            public static implicit operator double(JsonDynamicValue Value)
            {
                Verify.False(Value.IsString, "Value is a string.");
                return Value.GetDouble();
            }

            public static implicit operator JsonDynamicValue(bool Value)
            {
                return new JsonDynamicValue(Value ? "true" : "false", false);
            }

            public static implicit operator bool(JsonDynamicValue Value)
            {
                Verify.False(Value.IsString, "Value is a string.");
                return Value.GetBoolean();
            }

            public static bool operator ==(JsonDynamicValue L, JsonDynamicValue R)
            {
                return (L.Value == R.Value) & (L.IsString == R.IsString);
            }

            public static bool operator !=(JsonDynamicValue L, JsonDynamicValue R)
            {
                return !(L == R);
            }

            public override bool Equals(object obj)
            {
                var T = obj as JsonDynamicValue;
                if (T == null)
                    return false;
                return this == T;
            }

            public override int GetHashCode()
            {
                return Utilities.CombineHashCodes(this.Value.GetHashCode(), this.IsString.GetHashCode());
            }

            public override string ToString()
            {
                return (this.IsString ? "\"" : "") + this.Value + (this.IsString ? "\"" : "");
            }

            private readonly string _Value;

            public string Value
            {
                get
                {
                    return this._Value;
                }
            }

            private readonly bool _IsString;

            public bool IsString
            {
                get
                {
                    return this._IsString;
                }
            }

            internal const string True = "true";
            internal const string False = "false";
        }
    }
